Refactor Parabolic timer and add Parabolic smoother

dqyi11 opened the pull request:
(1) Refactor Parabolic timer into 
(1.a) convert an interpolated into a spline
(1.b) convert a spline into a timed spline
(2) Add Parabolic smoother 
(2.a) convert a spline to a dynamic path
(2.b) convert a dynamic path to a spine
(2.c) add HauserParabolicSmoother class that handles "doShortcut", "doBlend" and "doShortcutAndBlend"

mkoval commented on src/planner/parabolic/ParabolicUtil.hpp:
```diff
@@ -0,0 +1,72 @@
+/// \return a dynamic path
+std::unique_ptr<ParabolicRamp::DynamicPath>
+    convertToDynamicPath(const aikido::trajectory::Spline* _inputTrajectory,
+                         double& _startTime,
```
What is the purpose of this? :confused:

dqyi11 replied:
@mkoval A spline trajectory has a start time (not necessarily starting from 0). However, ParabolicRamp::DynamicPath always assumes starting from 0.
I did not modify the DynamicPath class. So I used a start time variable to save the information when converting.
When I convert the DynamicPath back to a Spline, I will use the start time to recover.

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -100,6 +100,18 @@ void Interpolated::evaluateDerivative(double _t, int _derivative,
     _tangentVector.setZero();
   }
 
+  //if(_t==getStartTime())
```
Nit: Remove dead code.

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -100,6 +100,18 @@ void Interpolated::evaluateDerivative(double _t, int _derivative,
   }
 
+  //if(_t==getStartTime())
+  if(_t==mWaypoints[0].t)
```
This is comparing two floating point numbers for exact equality, which is almost always a bug. What is the purpose of thsi check?

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -100,6 +100,18 @@ void Interpolated::evaluateDerivative(double _t, int _derivative,
+  if(_t==mWaypoints[0].t)
+  {
+      const auto segmentTime = mWaypoints[1].t - mWaypoints[0].t;
+      const auto alpha = (mWaypoints[1].t - mWaypoints[0].t) / segmentTime;
```
It looks like `alpha` will always be `1` because the numerator and denominator are both `mWaypoints[1].t - mWaypoints[0].t`. What am I missing/

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -100,6 +100,18 @@ void Interpolated::evaluateDerivative(double _t, int _derivative,
+  //if(_t==getStartTime())
+  if(_t==mWaypoints[0].t)
+  {
+      const auto segmentTime = mWaypoints[1].t - mWaypoints[0].t;
```
This will SEGFAULT if `mWaypoints.size() < 2`.

dqyi11 replied:
@mkoval  Could we assume that mWaypoints.size() is always >= 2 in an interpolated?

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -100,6 +100,18 @@ void Interpolated::evaluateDerivative(double _t, int _derivative,
+      mInterpolator->getDerivative(mWaypoints[0].state, mWaypoints[1].state,
+              _derivative, alpha, _tangentVector);
+      _tangentVector /= segmentTime;
+      return;
```
Why do we explicitly `return` here? :confused:

dqyi11 replied:
@mkoval I am taking _t == start_time as a specific case. So I will return here to skip running the other part of the function. This is an unfinished discussion. Do you still think we should support the derivative here?

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -152,6 +164,16 @@ const statespace::StateSpace::State* Interpolated::getWaypoint(
+{
+  if (_index < mWaypoints.size())
+    //return getStartTime() + mWaypoints[_index].t;
+    return mWaypoints[_index].t;
```
Won't this break if `getStartTime() != 0`?

dqyi11 replied:
In the Interpolated class, getStartTime() is taken as the time of the first way point.
"mWaypoints.front().t"
I will double check the correctness.

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -152,6 +164,16 @@ const statespace::StateSpace::State* Interpolated::getWaypoint(
+double Interpolated::getWaypointTime(size_t _index) const
+{
+  if (_index < mWaypoints.size())
+    //return getStartTime() + mWaypoints[_index].t;
```
If the current implementation is correct, remove this dead code.

mkoval replied:
@dqyi11 Please address this.

mkoval commented on src/trajectory/Spline.cpp:
```diff
@@ -198,5 +198,48 @@ Eigen::VectorXd Spline::evaluatePolynomial(
+void Spline::getWaypointDerivative(size_t _index, int _derivative,
+                                   Eigen::VectorXd& _tangentVector) const
+{
+  if (_index < getNumSegments())
```
This should be `getNumWapoints()`, not `getNumSegements()`.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+
+  std::shared_ptr<R2> mStateSpace;
+  Eigen::Vector2d mMaxVelocity;
+  Eigen::Vector2d mMaxAcceleration;
```
These are [fixed-size vectorizable types](https://eigen.tuxfamily.org/dox/group__TopicFixedSizeVectorizable.html), so you need to [annotate this class with `EIGEN_MAKE_ALIGNED_OPERATOR_NEW`](https://eigen.tuxfamily.org/dox/group__TopicStructHavingEigenMembers.html). Otherwise you will see runtime crashes on some architectures, namely `x86`.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+    auto interpolatedState = mStateSpace->createState();
+    Eigen::VectorXd splineVec, interpolatedVec;
+    Eigen::VectorXd splineTangent, interpolatedTangent;
+    for(double t=spline->getStartTime();t<=spline->getEndTime();t+=1e-2)
```
Use `StepSequence` to implement this stepping logic.

Also make `1e-2` a constant and decrease the value - perhaps to `1e-3`.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+        mStraightLine->evaluate(t, interpolatedState);
+        mStateSpace->logMap(splineState, splineVec);
+        mStateSpace->logMap(interpolatedState, interpolatedVec);
+        EXPECT_TRUE(interpolatedVec.isApprox(interpolatedVec));
```
Nit: Use `EXPECT_EIGEN_EQUAL` here instead of `EXPECT_TRUE`. It will generate better error messages.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+
+        spline->evaluateDerivative(t, 1, splineTangent);
+        mStraightLine->evaluateDerivative(t, 1, interpolatedTangent);
+        EXPECT_TRUE(splineTangent.isApprox(interpolatedTangent));
```
Nit: Also use `EXPECT_EIGEN_EQUAL` here.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+    }
+}
+
+TEST_F(ParabolicSmootherTests, doShort)
```
Typo: `doShortcut`.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+   state.setValue(p4);
+   nonStraightLine.addWaypoint(0.75, state);
+   state.setValue(p5);
+   nonStraightLine.addWaypoint(1., state);
```
Nit: Could we implement this test using only three waypoints?

dqyi11 replied:
No.... In Hauser's code, shortcut will be executed only when the waypoint number is larger than 3....

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+   smoothedTrajectory->evaluate(smoothedTrajectory->getEndTime(), state);
+   EXPECT_TRUE(p5.isApprox(state.getValue()));
+
+   EXPECT_TRUE(shortenDist <= orignDist);
```
Nit: Make this a strict inequality and increase the time limit until it succeeds with very high probability.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+
+   double originTime = nonStraightLine.getDuration();
+   double shortenTime = smoothedTrajectory->getDuration();
+   EXPECT_TRUE(shortenTime<=originTime);
```
Make this a strict inequality.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,240 @@
+
+   double originTime = nonStraightLine.getDuration();
+   double shortenTime = smoothedTrajectory->getDuration();
+   EXPECT_TRUE(shortenTime<=originTime);
```
See my comment above. Make this a strict inequality.

Also: There's a lot of duplicate code here. Can we move some of it to a helper function?

mkoval commented on src/planner/parabolic/ParabolicUtil.cpp:
```diff
@@ -0,0 +1,228 @@
+  auto outputPath = make_unique<ParabolicRamp::DynamicPath>();
+  // Apply the adjoint limits
+  outputPath->Init(toVector(_maxVelocity), toVector(_maxAcceleration));
+  outputPath->SetMilestones(milestones);
```
We need another version of this helper function - or an overload of this function - that optionally passes velocities to `SetMilestones`. We need that for all of the functions that accept non-piecewise linear splines.

mkoval commented on src/planner/parabolic/ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,150 @@
+
+  double startTime = 0.0;
+  auto dynamicPath = convertToDynamicPath(_inputTrajectory, startTime,
+                                          _maxVelocity, _maxAcceleration);
```
This call to `convertToDynamicPath` should preserve velocities.

mkoval commented on src/planner/parabolic/ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,150 @@
+                                            _blendIterations);
+  double startTime = 0.0;
+  auto dynamicPath = convertToDynamicPath(_inputTrajectory, startTime,
+                                          _maxVelocity, _maxAcceleration);
```
This call to `convertToDynamicPath` should preserve velocities.

mkoval commented on src/planner/parabolic/ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,150 @@
+
+  double startTime = 0.0;
+  auto dynamicPath = convertToDynamicPath(_inputTrajectory, startTime,
+                                          _maxVelocity, _maxAcceleration);
```
This call to `convertToDynamicPath` should preserve velocities.

dqyi11 commented on include/aikido/planner/parabolic/ParabolicTimer.hpp:
```diff
@@ -34,6 +34,37 @@ std::unique_ptr<trajectory::Spline> computeParabolicTiming(
+      const Eigen::VectorXd& _maxVelocity,
+      const Eigen::VectorXd& _maxAcceleration);
+
+/// Convert an interpolated trajectory to a spline trajectory
```
@mkoval  The velocities are used in conversion. Do you think it is ignored?

dqyi11 commented on src/planner/parabolic/ParabolicTimer.cpp:
```diff
@@ -154,97 +105,26 @@ std::unique_ptr<trajectory::Spline> computeParabolicTiming(
-    positionPrev = positionCurr;
-    velocityPrev = velocityCurr;
-  }
+  double startTime = 0.0;
```
Because in converting to a dynamic path, the start time information is lost.
I do it this way to remind a user that he needs to save the start time if he wants to convert the dynamic path back to a spline.

dqyi11 commented on src/planner/parabolic/ParabolicUtil.cpp:
```diff
@@ -0,0 +1,228 @@
+  }
+
+  auto outputPath = make_unique<ParabolicRamp::DynamicPath>();
+  // Apply the adjoint limits
```
Yes, I am going to add velocities.

dqyi11 commented:
@mkoval  All the comments are addressed.
The only concern is whether we should support the derivative at the start time of an interpolated trajectory, which defines the modifications in the interploated class.
I think we need the velocity to convert an interpolated trajectory to a spline trajectory, because a spline trajectory assumes the smoothness.
Please let me know what you think about it. Thanks!

dqyi11 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+}
+
+bool tryBlend(ParabolicRamp::DynamicPath& dynamicPath,
+              ParabolicRamp::RampFeasibilityChecker feasibilityChecker,
```
Only changed it to ParabolicRamp::RampFeasibilityChecker&. 
Because it calls ```dynamicPath.TryShortcut(t1, t2, feasibilityChecker);``` , which is ``` bool TryShortcut(Real t1,Real t2,RampFeasibilityChecker& check); ```

dqyi11 commented on include/aikido/planner/parabolic/ParabolicSmoother.hpp:
```diff
@@ -0,0 +1,124 @@
+///
+/// while _timelimit is not out:
+/// - Randomly sample two times uniformly at random.
+/// - Find two waypoints according two times in _inputTrajectory.
```
@mkoval   Should I then say "points" instead of "waypoints"?

dqyi11 commented:
I have addressed all the comments.

The only thing left is whether we shall keep "useVelocity" in converting an interpolated to a spline.
This can be useful in the ParabolicSmoother.

In that, I convert a spline trajectory to a dynamic path without providing velocity.
Then dynamic path will re-evaluate the velocity.
If velocities are given, dynamic path will not re-evaluate the velocities.
Thus no new timing information will be added.
Does this make sense?

jslee02 commented on include/aikido/planner/parabolic/ParabolicTimer.hpp:
```diff
@@ -30,17 +30,18 @@ namespace parabolic {
 /// Convert an interpolated trajectory to a spline trajectory
-/// This function requires the \c _inputTrajectory to use a \c GeodesicInterpolator.
+/// This function requires the \c _inputTrajectory to use a \c
+/// GeodesicInterpolator.
```
Nit: Could you fix the wrapping? ClangFormat isn't good at it. :disappointed:

jslee02 replied:
Nit: I thought "So the conversion ..." was the continuing sentence without line breaking.

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+  {
+  }
+
+  virtual bool ConfigFeasible(ParabolicRamp::Vector const& x)
```
Nit: Please use `const X&` pattern rather than `X const&`.

jslee02 replied:
Also, please begin with a lower letter for member functions.

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+    , statespace_(testable_->getStateSpace())
+    , interpolator_(statespace_)
+  {
+  }
```
Nit: Please add `// Do nothing` comment to indicate that this function is intentionally empty (but initializing the member variables).

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+  }
+
+  virtual bool SegmentFeasible(
+      ParabolicRamp::Vector const& a, ParabolicRamp::Vector const& b)
```
Nit: Ditto

jslee02 replied:
Please rename to `segmentFeasible(...)`.

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
-    double                                   checkResolution_;
-    aikido::statespace::StateSpacePtr        statespace_;
-    aikido::statespace::GeodesicInterpolator interpolator_;
+  aikido::constraint::TestablePtr testable_;
```
Nit: Could you rename these private member variables using the pattern of `mX`? I believe we don't use the trailing underscore for member variables (even for private member variables).

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
 };
 
-bool needsBlend(ParabolicRamp::ParabolicRampND const &rampNd)
+bool needsBlend(ParabolicRamp::ParabolicRampND const& rampNd)
```
Nit: `const ParabolicRamp::ParabolicRampND&`

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+  // blending can completely remove waypoints from the trajectory in the case
+  // that two waypoints are closer than _blendRadius together - which means
+  // that waypoint indicies can change between iterations of the algorithm.
+  size_t const numRamps = dynamicPath.ramps.size();
```
Nit: `const size_t`.

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+  // that two waypoints are closer than _blendRadius together - which means
+  // that waypoint indicies can change between iterations of the algorithm.
+  size_t const numRamps = dynamicPath.ramps.size();
+  double const tMax = dynamicPath.GetTotalTime();
```
Nit: `const double`

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
-                rampNd.blendAttempts++;
-            }
-        }
+      double const t1 = std::max(t - dtShortcut, -ParabolicRamp::EpsilonT);
```
Nit: `const double`

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
-            }
-        }
+      double const t1 = std::max(t - dtShortcut, -ParabolicRamp::EpsilonT);
+      double const t2
```
Nit: `const double`

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+      double const t2
+          = std::min(t + dtShortcut, tMax + ParabolicRamp::EpsilonT);
+
+      bool const success = dynamicPath.TryShortcut(t1, t2, feasibilityChecker);
```
Nit: `const bool`

jslee02 commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -13,172 +13,184 @@ namespace planner {
+  for (int attempt = 0; attempt < blendIterations; ++attempt)
+  {
+    while (tryBlend(dynamicPath, feasibilityChecker, attempt, dtShortcut))
+      ;
```
:scream: What's happening here?

jslee02 replied:
Nit: At first, I thought this was a mistake, but now I got it. I would like to suggest make this code easier to read something like:
```cpp
bool success;
do 
{
  success = tryBlend(...);
} while (success);
```
Also, I'm not sure if which one is logically correct between `while(success)` and `while(!success)`. :confused:

mkoval replied:
The current logic is correct. The core issue is that blending can completely remove waypoints from the trajectory, which makes bookkeeping via the time parameter extraordinarily difficult. We dodge that issue by allowing `tryBlend` to always start at the beginning of the trajectory. We continue to call `tryBlend` in a loop until it finds nothing left to blend.

jslee02 replied:
That makes sense. It would be nice to add this as a comment at least to `tryBlend()`. 

I still prefer to change the code to be self-documenting. For example:
```cpp
bool needMoreBlending;
do 
{
  needMoreBlending = tryBlend(...);
} while (needMoreBlending);
```

mkoval commented on include/aikido/planner/parabolic/ParabolicSmoother.hpp:
```diff
@@ -0,0 +1,137 @@
+    aikido::constraint::TestablePtr _feasibilityCheck,
+    const Eigen::VectorXd& _maxVelocity,
+    const Eigen::VectorXd& _maxAcceleration,
+    aikido::util::RNG& _rngSeed,
```
Nit: This *is* the RNG, not a seed for the RNG. 😉

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+    mStateSpace->expMap(eigA, startState);
+    mStateSpace->expMap(eigB, goalState);
+
+    aikido::util::VanDerCorput vdc{1, false, false, mCheckResolution};
```
Nit: Add a comment explaining why `includeStartpoint` and `includeEndpoint` can be `false`.

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+    aikido::util::RNG& rng)
+{
+  if (timelimit < 0.0)
+    throw std::runtime_error("Timelimit should be non-negative");
```
Nits:
- Add a bounds check for `checkResolution`.
- Use `std::invalid_argument` instead of `std::runtime_error` for these exceptions.

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+  SmootherFeasibilityCheckerBase base(testable, checkResolution);
+  ParabolicRamp::RampFeasibilityChecker feasibilityChecker(&base, tolerance);
+
+  std::chrono::time_point<std::chrono::system_clock> startTime, currentTime;
```
Nits:
- `currentTime` is only used inside the loop, so move its definition to where it is first used.
- Initialize `startTime` on the same line where the variable it is defined.

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+                      currentTime - startTime)
+                      .count();
+  }
+  return true;
```
This `bool` seems to be meaningless. We should check whether `TryShortcut` ever reported success, i.e. returned `true`, and report that here.

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+  if (blendIterations <= 0)
+    throw std::runtime_error("Blend iterations should be positive");
+  if (blendRadius <= 0.0)
+    throw std::runtime_error("Blend radius should be positive");
```
Nits:
- Add bounds checks for `checkResolution` and `tolerance`.

mkoval commented on src/planner/parabolic/ParabolicTimer.cpp:
```diff
@@ -135,6 +39,50 @@ std::unique_ptr<trajectory::Spline> computeParabolicTiming(
+    stateSpace->logMap(nextState, nextVec);
+
+    _inputTrajectory.evaluateDerivative(currentTime, 1, currentVelocity);
+    _inputTrajectory.evaluateDerivative(nextTime, 1, nextVelocity);
```
This is not meaningful.

It is not possible to convert an `Interpolated` to a `Spline` in the general case - it is only possible when the `Interpolator` used by the input evaluates a polynomial (i.e. is a spline itself). In practice, the only case we need to handle is the `GeodesicInterpolator`, which is equivalent to a piecewise linear spline.

A piecewise linear spline is not C1 continuous, ,so `evaluateDerivative` is undefined at waypoints. Since that is the case, I suggest:
1. verifying that `_inputTrajectory->getInterpolator()` is a `GeodesicInterpolator`
2. creating a `LinearSplineProblem` instead of a `CubicSplineProblem`
3. removing the derivative `addConstantConstraint(_, 1, _)` calls.

mkoval commented on src/planner/parabolic/ParabolicTimer.cpp:
```diff
@@ -154,94 +102,12 @@ std::unique_ptr<trajectory::Spline> computeParabolicTiming(
-  }
+  double startTime = _inputTrajectory.getStartTime();
+  auto dynamicPath = convertToDynamicPath(
+      _inputTrajectory, _maxVelocity, _maxAcceleration, false);
```
This version should pass `_useVelocity = true`!

We can convert two types of inputs to a `DynamicPath`:

1. Untimed piecewise linear `Interpolated` trajectories that are untimed and are converted to timed `Spline` trajectories that stop at each waypoint [which is necessary to follow a piecewise linear trajectory under acceleration bounds]. In this case, `_useVelocity = false`.
2. Timed parabolic `Spline` trajectories that we want to shortcut or blend. In this case, it is important that we preserve the exact geometry of the input path, so `it is important that `_useVelocity = true`. Otherwise, `DyanmicRamp` will treat the `Spline`'s knot points as waypoints connected by piecewise linear segments.

dqyi11 replied:
@mkoval By what you suggested, here we should pass `_useVelocity = false`, right? Because we are converting an untimed piecewise linear `Spline` to a `DynamicPath` (after making changes by above suggestion).

mkoval replied:
Yes, that is correct. :+1: 

Although, as per our discussion on Slack, we could  also remove this function entirely. We only really need (1) to support untimed `Interpolated` inputs and (2) to support timed `Spline` inputs.

dqyi11 replied:
As discussed with @mkoval , in `ParabolicTimer', there should be 
(1) `computeParabolicTime()` that converts an untimed piecewise linear `Interplotaed` to a timed parabolic cubic `Spline`; and
(2) `convertToSpline` that converts an untimed piecewise linear `Interploated` to a untimed piecewise linear `Spline` (Not used now).

mkoval commented on src/planner/parabolic/ParabolicUtil.cpp:
```diff
@@ -0,0 +1,230 @@
+namespace aikido {
+namespace planner {
+namespace parabolic {
+
```
Nit: These helper functions belong in a `detail` namespace.

mkoval commented on src/planner/parabolic/ParabolicUtil.cpp:
```diff
@@ -0,0 +1,230 @@
+    milestones.emplace_back(toVector(currVec));
+
+    _inputTrajectory.getWaypointDerivative(iwaypoint, 1, tangentVector);
+    velocities.emplace_back(toVector(tangentVector));
```
Nit: Don't call `getWaypointDerivative` if `_useVelocity` is `false`.

mkoval replied:
Resolved by splitting the `Spline` and `Interpolated` cases into separate overloads. 👍

mkoval commented on src/planner/parabolic/ParabolicUtil.hpp:
```diff
@@ -0,0 +1,70 @@
+    convertToDynamicPath(const aikido::trajectory::Spline& _inputTrajectory,
+                         const Eigen::VectorXd& _maxVelocity,
+                         const Eigen::VectorXd& _maxAcceleration,
+                         bool _useVelocity = true);
```
@dqyi11 I would still like to address this.

mkoval commented on src/trajectory/Interpolated.cpp:
```diff
@@ -100,6 +100,17 @@ void Interpolated::evaluateDerivative(double _t, int _derivative,
     _tangentVector.setZero();
   }
 
+  if(_t==mWaypoints[0].t)
```
We're testing two floating point numbers of equality here. This is usually a bug, because the numbers may differ due to numerical precision issues. Please either add a tolerance or add a comment about why exact equality is appropriate here (e.g. because `t` was copied out of `mWaypoints[0].t`).

dqyi11 replied:
I'd prefer to add a comment here. Because it is only when _t is exactly the same with mWaypoints[0].t , we do something special. Also, @mkoval do you think we shall not support evaluate Derivative when _t < mWaypoints[0].t?

mkoval replied:
> Because it is only when _t is exactly the same with mWaypoints[0].t , we do something special.

I still don't really understand why we're checking for exact equality. What if, due to numerical precision, we end up getting `mWaypoints[0].t - eps`? :confused: 

> Also, @mkoval do you think we shall not support evaluate Derivative when _t < mWaypoints[0].t?

I think we should support this. We had a long discussion about this last year and decided that we would not implement hard bounds checks in Aikido arrays, unless the underlying implementation absolutely requires it. The issue is that numerical precision can result in an algorithm querying the result slightly outside of the nominal bounds.

This numerical imprecision is often unavoidable. It's often more useful to return a reasonable value, e.g. by extrapolating the last segment or extending the trajectory with a constant segment than to report an error. @psigen and I got burned by this repeatedly when implementing `or_parabolicsmoother` - as you can see by the variety of `epsilon`s. :sweat_smile:

dqyi11 replied:
@mkoval When _t == mWaypoints[0].t, `int idx == getWaypointIndexAfterTime(_t)` returns 0 to idx. Then a domain error will be thrown. I needed the velocity of the start time of an `Interploated` to convert it to a parabolic cubic `Spline`. Since you suggested to convert it into a piecewise linear `Spline`, I think I do not need the velocities any more. 

Since we are not currently supporting the time before the start time now. As you pointed out previously that it is not continuous here thus no derivative shall be supported. Maybe I could just revert all the changes made here. Does this make sense?

mkoval replied:
I am not 100% sure that I follow. For (1), the input `Interpolated` is piecewise linear so we do not need any velocities [as you noted]. If that is consistent with your understanding, then I agree that we should revert those changes.

dqyi11 replied:
OK. I will revert it.

mkoval commented on src/trajectory/Spline.cpp:
```diff
@@ -198,5 +198,48 @@ Eigen::VectorXd Spline::evaluatePolynomial(
+
+//=============================================================================
+double Spline::getWaypointTime(size_t _index) const
+{
```
This is resolved.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,231 @@
+protected:
+  void SetUp() override
+  {
+    mRng = aikido::util::RNGWrapper<std::mt19937>( std::random_device{}() );
```
Please use a deterministic seed (e.g. `0`) in unit tests so their result is not stochastic.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,231 @@
+    state.setValue(Vector2d(3., 4.));
+    mStraightLine->addWaypoint(1., state);
+
+    mNonStraightLineLength = initNonStraightLine();
```
It's odd to have one output (`mNonStraightLine`) in a member variable and a second `mNonStraightLineLength` as a return value. Please stick to one convention or the other for this function.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,231 @@
+    auto currState = mStateSpace->createState();
+    auto nextState = mStateSpace->createState();
+    Eigen::VectorXd currVec, nextVec;
+    for(size_t i=0; i<spline->getNumWaypoints()-1;++i)
```
Nit: This will overflow if `getNumWaypoints() == 0`, which could cause the test to SEGFAULT instead of fail properly. Please refactor the loop so that is not an issue, e.g. by starting with `i = 1` and replacing `i` with `i - 1` in the body.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,231 @@
+
+TEST_F(ParabolicSmootherTests, convertInterpolatedToSpline)
+{
+    auto spline = convertToSpline(*mStraightLine.get());
```
Nit: `*mStraightLine.get()` is equivalent to `*mStraightLine`.

mkoval commented on tests/planner/parabolic/test_ParabolicSmoother.cpp:
```diff
@@ -0,0 +1,231 @@
+        mStraightLine->evaluateDerivative(t, 1, interpolatedTangent);
+        EXPECT_EIGEN_EQUAL(splineTangent, interpolatedTangent, mTolerance);
+    }
+}
```
Add tests for:
- More than one segment, e.g. the non-striaghtline trajectory.
- A trajectory with a non-zero start time.

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,216 @@
+
+  // tryBlend always starts at the beginning of the trajectory.
+  // Without this bookkeeping, tryBlend would could any blend that fails
+  // multiple times.
```
Nit: Typo "would could." I think this should say "tryBlend would retry any blend that fails multiple times."

mkoval commented on src/planner/parabolic/HauserParabolicSmootherHelpers.cpp:
```diff
@@ -0,0 +1,216 @@
+    {
+      noMoreBlending
+          = tryBlend(dynamicPath, feasibilityChecker, attempt, dtShortcut);
+    } while (noMoreBlending);
```
Nit: The name `noMoreBlending` implies the opposite of what it actually is. I suggest renaming the flag to `needsMoreBlending`.

mkoval commented on src/planner/parabolic/ParabolicUtil.cpp:
```diff
@@ -0,0 +1,228 @@
+  else if (dynamic_cast<const R<6>*>(_stateSpace) != nullptr)
+  {
+    return true;
+  }
```
Resolved.
